minorminer/utils/feasibility.py

```diff
@@ -0,0 +1,222 @@
+# Copyright 2024 D-Wave
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""Determining embedding feasibility is NP-complete.
+Efficient methods and heuristics are collected together in this module,
+that provide necessary conditions and bounds on feasibility.
+This includes determination of a lower bound on the (square)
+lattice size required for chimera, pegasus and zephyr dwave_networkx graph
+types, and a feasibility filter for general graphs based on the degree
+distributions of target and source graphs.
+"""
+import dwave_networkx as dnx
+import networkx as nx
+import numpy as np
+from collections import Counter
+
+
+def embedding_feasibility_filter(
+    S: nx.Graph, T: nx.Graph, one_to_one: bool = False
+) -> bool:
+    """Feasibility filter for embedding.
+
+    Determines if the source graph `S` can be subgraph-embedded onto the target
+    graph `T` based on their degree distributions. Returns `False` if embedding
+    is definitely infeasible; otherwise, returns `True`. False positives are
+    permissible because deciding the graph isomorphism problem is NP-complete,
+    and this filter is designed to be efficient.
+
+    The degree distribution test is a heuristic; stronger filters are possible
+    by exploiting additional graph structure.
+
+    Args:
+        S: The source graph to embed.
+        T: The target graph in which to embed.
+        one_to_one: If True, only 1-to-1 (subgraph) embeddings are allowed.
+            Defaults to False, permitting minor embeddings.
+
+    Returns:
+        bool: `False` if subgraph embedding is definitely infeasible, `True`
+            otherwise.
```

Comment:
No indent on returns.
```suggestion
        bool: `False` if subgraph embedding is definitely infeasible, `True`
        otherwise.
```